cloudmarker/util.py

```diff
@@ -453,6 +455,118 @@ def pluralize(count, word, *suffixes):
+    unspecified, i.e., if host or port are ``''`` and/or ``0``,
+    respectively, the OS default behavior is used. See
+    :class:`smtplib.SMTP` documentation for more details on this.
+
+    We recommend these parameter values:
+
+    - Leave ``ssl_mode`` unspecified (thus ``'ssl'`` by default) if
+      your SMTP server supports SSL.
+
+    - Set ``ssl_mode`` to ``'starttls'`` explicitly if your SMTP server
+      does not support SSL but it supports STARTTLS.
+
+    - Set ``ssl_mode`` to ``'disable'`` explicitly if your SMTP server
+      supports neither SSL nor STARTTLS.
+
+    - Set ``host`` to the SMTP hostname or address explicitly.
+
+    - Leave ``port`` unspecified (thus ``0`` by default), so that the
+      appropriate port is chosen automatically.
+
+    With these recommendations, this function should do the right thing
+    automatically, i.e., connect to port 465 if ``use_ssl`` is
+    unspecified or ``False`` and port 25 if ``use_ssl`` is ``True``.
+
+    Note that in case of SMTP, there are two different encryption
+    protocols in use:
+
+    - SSL/TLS (or implicit SSL/TLS): SSL/TLS is used from the beginning
+      of the connection. This occurs typically on port 465. This is
+      enabled by default (``ssl_mode`` as ``'ssl'``).
+
+    - STARTTLS (or explicit SSL/TLS): The SMTP session begins as a
+      plaintext session. Then the client (this function in this case)
+      makes an explicit request to switch to SSL/TLS by sending the
+      ``STARTTLS`` command to the server. This occurs typically on port
+      25 or port 587. Set ``ssl_mode`` to ``'starttls'`` to enable this
+      behaviour
+
+    If ``username`` is unspecified or specified as an empty string, no
+    SMTP authentication is done. If ``username`` is specified as a
+    non-empty string, then SMTP authentication is done.
+
+    Arguments:
+        from_addr (str): Sender's email address.
+        to_addrs (list): A list of :obj:`str` objects where each
+            :obj:`str` object is a recipient's email address.
+        subject (str): Email subject.
+        content (str): Email content.
+        host (str): SMTP host.
+        port (int): SMTP port.
+        ssl_mode (str): SSL mode to use: ``'ssl'`` for SSL/TLS
+            connection (the default), ``'starttls'`` for STARTTLS, and
+            ``'disable'`` to disable SSL.
+        username (str): SMTP username.
+        password (str): SMTP password.
+        debug (int or bool): Debug level to pass to
+            :meth:`SMTP.set_debuglevel` to debug an SMTP session. Set to
+            ``0`` (the default) or ``False`` to disable debugging. Set
+            to ``1`` or ``True`` to see SMTP messages. Set to ``2`` to
+            see timestamped SMTP messages.
+    """
+    log_data = ('from_addr: {}; to_addrs: {}; subject: {}; host: {}; '
+                'port: {}; ssl_mode: {}'
+                .format(from_addr, to_addrs, subject, host, port, ssl_mode))
+    try:
+        if ssl_mode == 'ssl':
```

Comment:
Can we define constants for these `ssl_modes` rather than hardcoding it.

Reply:
What would be the benefit of adding constants for these? I don't see any benefit because we are not reusing this string anywhere else. I do see a disadvantage: adding an additional level of indirection. There needs to be some justification for adding an additional level of indirection like this. Do you see any justification for it?

Reply:
It would be easier for the plugin developers who would be using `_send_email` would have a list of `ssl_mode` options. Though the docstring gives a very crystal clear picture of the ssl mode, providing constants will give an easy options to show all the ssl_mode available.

Reply:
If I understand your point correctly, the rationale here is that having the constants together at some place, say, `util.SSL`, `util.STARTTLS`, and `util.DISABLED` is going to be easier for plugin developers. We could choose another module say `constants` instead of `util` to keep the constants; that is beside the point.

I am not convinced about this point though. In fact, the opposite seems to be true to me. Let us imagine a plugin developer story for a while:

As a plugin developer, I would go to the `cloudmarker.util` documentation on [Read The Docs: cloudmarker.util](https://cloudmarker.readthedocs.io/en/latest/api/cloudmarker.html#module-cloudmarker.util). I will jump to the documentation for `send_email()`. Now I would read something like this:

> ssl_mode (str): SSL mode to use: `util.SSL` for SSL/TLS connection (the default), `util.STARTTLS` for STARTTLS, and `util.DISABLE` to disable SSL.

Now instead of the actual strings I care about, I see some constants. It gives me little idea of what the actual strings look should look like in the configuration. Granted I can just click on `util.STARTTLS` to jump to the docstring of `util.STARTTLS` and see that it is defined as `'starttls'`.

But what did we achieve with this? I don't think we achieved ease of use. It appears to me that we just increased a level of indirection, interrupted the reader, made him/her jump to another section of the documentation unnecessarily when he/she was much rather interested in reading the `util.send_email()` documentation, and increased the number of lines of code.

The current documentation reads like this:

> ssl_mode (str): SSL mode to use: `'ssl'` for SSL/TLS connection (the default), `'starttls'` for STARTTLS, and `'disable'` to disable SSL.

I find this much more developer and reader friendly. It also tells the user instantly what to put in the configuration file.